Moving construction of IR into runtime parser project

JaroslavTulach opened the pull request:
### Pull Request Description

Moving and shuffling around to isolate construction of `IR` from the rest of the `runtime` project. All the classes can be found in `org.enso.compiler.core` and subpackages.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- All code has been tested:
  - [x] Unit tests have been written where possible and continue to pass

JaroslavTulach committed:
Given the scala mime type a position

JaroslavTulach committed:
Locking the transitive NPM dependencies

JaroslavTulach committed:
Skeleton of outline view and completion provider for VSCode

JaroslavTulach committed:
runtime-parser project & further changes to encapsulate IR creation

JaroslavTulach committed:
Merging with most recent develop branch

JaroslavTulach committed:
Introducing parser's ProcessingPass and making runtime.IRPass extend it

JaroslavTulach committed:
Moving parser into org.enso.compiler.core

JaroslavTulach committed:
Removing garbage

JaroslavTulach commented on engine/runtime-instrument-common/src/main/scala/org/enso/interpreter/instrument/job/EnsureCompiledJob.scala:
```diff
@@ -289,9 +289,10 @@ final class EnsureCompiledJob(protected val files: Iterable[File])
+    changeset.getClass()
+    source.getClass()
+    ctx.getClass()
+    val scopeIds = Seq()
```
TBD: We will need to do something with this...

JaroslavTulach replied:
Metadata parser is back since [b6513fb](https://github.com/enso-org/enso/pull/7096/commits/b6513fb8ceae28846e33961bab263226ceaea382)

radeusgd commented on engine/runtime-parser/src/main/java/org/enso/compiler/core/ConstantsNames.java:
```diff
@@ -0,0 +1,8 @@
+  public static final String SELF_TYPE_ARGUMENT = "Self";
+  public static final String THAT_ARGUMENT = "that";
+  public static final String FROM_MEMBER = "from";
+}
```
Please make sure that all files contain a newline at EOF, we try to keep this convention I think.

JaroslavTulach replied:
There is an empty 9th line.

JaroslavTulach replied:
I see an empty 9th line: 
![obrazek](https://github.com/enso-org/enso/assets/26887752/e9332b41-c90f-4795-ae28-fea9ebaae2d0)

radeusgd replied:
![image](https://github.com/enso-org/enso/assets/1436948/76d76486-ea58-4bb1-b765-881c5fc5e146)

GitHub shows that the newline is missing (see the red 'stop sign' at the end). 

It is indeed weird that your editor shows the newline if it seems to be missing in git. Could it be some misconfiguration?

JaroslavTulach replied:
Will be addressed by https://github.com/enso-org/enso/pull/7115/commits/ef9ee738fe729aeb4126514098d01dd60ac3e35c

radeusgd reviewed (approved):
So great that the parser gets a separate module, this looks so much cleaner 🎉

radeusgd commented:
Note that you may need to run `gatherLicenses` - while the dependencies did not change, the internal structure of dependencies has changed - so the hash that is used to verify the dependency structure may need to be updated (but no other action should be needed if dependencies are not moved around).

hubertp reviewed (changes requested):
I'm not a fan of all of those `Any` instances and I think this can be removed by even deeper reorganization. This might take a while though. Also currently the project does not compiler in `ChangesetBuilder`

hubertp committed:
Fix compilation, avoid Any casting

Tweaks that avoid passing around `Compiler` as `Any`. We could probably
improve casting in `MetadataStorage` but it is slightly more involving.

hubertp reviewed (commented):
@JaroslavTulach Added some changes that addressed my concerns. Let me know if this is fine with you.

Akirathan reviewed (approved):
I like that `EnsoCompiler` is renamed to `EnsoParser`. Finally, there won't be both `EnsoCompiler` and `Compiler`.

JaroslavTulach commented on engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/ProcessingPass.scala:
```diff
@@ -10,6 +12,8 @@ trait ProcessingPass extends Serializable {
 object ProcessingPass {
   trait Metadata extends Serializable {
 
+    type Compiler <: CompilerStub
```
Interesting. Thank you Hubert.

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/pass/IRPass.scala:
```diff
@@ -115,52 +114,14 @@ object IRPass {
-      * @param compiler the Enso compiler
-      * @return `this`, but prepared for serialization
-      */
-    def prepareForSerialization(compiler: Compiler): Metadata
```
By defining `type Compiler` we no longer need to override these methods. That's great.

JaroslavTulach committed:
Removing no longer needed licenses

JaroslavTulach committed:
scalafmtSbt

JaroslavTulach committed:
Returning back metadata parser
